### ISSUE TITLE
Strange statistics of one_of strategy output

### ISSUE THREAD

**Issue by Tiendil** [HypothesisWorks/hypothesis#3193]
Hi!

I did some experiments with hypothesis (great library!) and found that the first element of ``one_of`` strategy generated statistically more often than other elements when the total elements count is not equal to a power of 2.

I expect, that elements distribution always should be near-uniform distribution.

Here is the example code:

```python
from collections import Counter

from hypothesis.strategies import just, one_of

k = 10
n = 1000

elements = [just(i) for i in range(k)]

for i in range(1, k + 1):
    strategy = one_of(elements[:i])
    counts = Counter(strategy.example() for _ in range(n))
    print(f'distribution for {i} elements: {counts}')
```

And output

```
distribution for 1 elements: Counter({0: 1000})
distribution for 2 elements: Counter({0: 523, 1: 477})
distribution for 3 elements: Counter({0: 510, 2: 250, 1: 240})
distribution for 4 elements: Counter({1: 257, 2: 253, 0: 245, 3: 245})
distribution for 5 elements: Counter({0: 523, 1: 123, 4: 123, 3: 120, 2: 111})
distribution for 6 elements: Counter({0: 355, 4: 142, 3: 137, 5: 128, 1: 125, 2: 113})
distribution for 7 elements: Counter({0: 268, 6: 140, 2: 125, 5: 123, 4: 116, 3: 115, 1: 113})
distribution for 8 elements: Counter({0: 140, 5: 129, 1: 125, 3: 124, 4: 124, 2: 124, 6: 118, 7: 116})
distribution for 9 elements: Counter({0: 523, 8: 70, 5: 63, 4: 63, 3: 60, 6: 59, 2: 57, 1: 56, 7: 49})
distribution for 10 elements: Counter({0: 472, 5: 70, 6: 68, 9: 65, 8: 60, 4: 58, 7: 57, 2: 53, 1: 51, 3: 46})
```

Look at the count of "0" in comparison to other elements.

Also reproduced with ``given``

```python
from collections import Counter

from hypothesis import given
from hypothesis.strategies import just, one_of

k = 10
n = 100

elements = [just(i) for i in range(k)]

for i in range(1, k + 1):
    strategy = one_of(elements[:i])

    counts = Counter()

    @given(strategy)
    def test(x):
        counts.update([x])

    for _ in range(n):
        test()

    print(f'distribution for {i} elements: {counts}')
```

Output

```
distribution for 1 elements: Counter({0: 100})
distribution for 2 elements: Counter({0: 100, 1: 100})
distribution for 3 elements: Counter({0: 200, 2: 100, 1: 100})
distribution for 4 elements: Counter({0: 100, 2: 100, 3: 100, 1: 100})
distribution for 5 elements: Counter({0: 400, 3: 100, 1: 100, 2: 100, 4: 100})
distribution for 6 elements: Counter({0: 300, 5: 100, 3: 100, 1: 100, 2: 100, 4: 100})
distribution for 7 elements: Counter({0: 200, 2: 100, 3: 100, 1: 100, 4: 100, 5: 100, 6: 100})
distribution for 8 elements: Counter({0: 100, 2: 100, 7: 100, 4: 100, 5: 100, 3: 100, 6: 100, 1: 100})
distribution for 9 elements: Counter({0: 604, 2: 129, 8: 127, 5: 126, 7: 126, 4: 125, 1: 124, 6: 122, 3: 117})
distribution for 10 elements: Counter({0: 522, 6: 125, 4: 123, 5: 122, 8: 120, 7: 120, 9: 118, 2: 118, 1: 117, 3: 115})
```

**Hypothesis version:** 6.31.3

**Comment by Zac-HD** [comment]
Hey @Tiendil - thanks for the writeup!  In this case it's basically "working as intended"; this pattern is due to the interaction between a couple of different features:

- Our underlying choice is roughly uniform-at-random; we draw random bits and use rejection sampling if we don't have a power-of-two size.
- *But also*, the bits aren't actually random!  We have a huge pile of heuristics and structure-inducing tricks in `hypothesis.internal.conjecture` which help to find bugs... though they make reasoning about the distribution harder too.

In this case, I'd guess that rejection sampling for non-power-of-two sizes is making it more likely that we reuse a previous value, and "all zeros" is our usual starting point - so that's more common even before our "use more zeros" heuristic.

Bottom line, we don't tend to think of this kind of thing as a bug, and it tends to wash out much more as you scale up to more complicated data.  Please do let us know if you think we might fail to generate something ~ever though; cases like #3170 are often bugs!
